[IMP] mail, web: call the installation prompt from the notification area

Iucapad opened the pull request:
The NotificationItem component has been modified to support a new slot allowing to display some content at the right of the notification item. It is used in this commit to display an additional 'Install button', inviting the user to click on the notification. Note that the entire notification is still clickable, it is only a visual indicator.

Once the notification is clicked, the installation process is triggered:

- In case the system supports the native installation prompt, we now display it, and the user is free to install the PWA from there.

- On Apple devices, a custom popup^is displayed to the user, explaining the steps to install the website as an application, since iOS and macOS don't support a native prompt, and won't be using one anytime soon, since it's a design decision from their side.

task-3275275

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/135866).

alexkuhn commented on addons/mail/static/src/core/web/messaging_menu.js:
```diff
@@ -71,6 +74,14 @@ export class MessagingMenu extends Component {
         return tab === "chat" ? ["chat", "group"] : [tab];
     }
 
+    get canBeInstalled() {
+        return (
+            this.installPrompt.canPromptToInstall &&
+            !isDisplayStandalone() &&
+            browser.location.pathname !== "/web/tests"
+        );
+    }
```
Could this be moved to `installPrompt` service?
The `isDisplayStandalone()` and `browser.location.pathname !== "/web/tests"` does not seem to be conceptually specific to messaging menu

Iucapad replied:
No problem, I will move it to the service directly

alexkuhn commented on addons/mail/static/src/core/web/notification_item.js:
```diff
@@ -20,6 +20,7 @@ export class NotificationItem extends Component {
         "onClick",
         "onSwipeLeft?",
         "onSwipeRight?",
+        "requestTitle?",
```
Looks unused

Iucapad replied:
Oh you’re right. I’ll remove it

pparidans commented on addons/web/static/src/core/install_prompt/install_prompt.scss:
```diff
@@ -0,0 +1,30 @@
+    height: unset !important;
+	-webkit-backdrop-filter: blur(5px);
+	backdrop-filter: blur(5px);
+	background: rgba(#FFF, .7);
```
About the background's color, what happens in dark mode ? Could you use an existing variable instead (at least for the "#FFF") ?

pparidans commented on addons/web/static/src/core/install_prompt/install_prompt.scss:
```diff
@@ -0,0 +1,30 @@
+            background: none !important;
+
+            button {
+                color: black !important;
```
Same about the color

pparidans commented on addons/web/static/src/core/install_prompt/install_prompt.scss:
```diff
@@ -0,0 +1,30 @@
+
+    &.mobile {
+        width: 90% !important;
+        inset: auto auto 5px 50% !important;
```
Why "5px" ?

Iucapad replied:
So the dialog is not touching the borders of the screen. I guess I may instead use margins

pparidans commented on addons/web/static/src/core/install_prompt/install_prompt.scss:
```diff
@@ -0,0 +1,30 @@
+
+    &.desktop {
+        width: fit-content !important;
+        inset: 5px auto auto 5px !important;
```
Same about the size

pparidans commented on addons/web/static/src/core/install_prompt/install_prompt.scss:
```diff
@@ -0,0 +1,30 @@
+.o_install_prompt {
+    border-radius: 1em;
```
Use class `rounded-X` instead ?

Iucapad replied:
I used this to use an "appleish" look, to make it look like a native popup.
We only have up to rounded-3, which is more squared and looks weird to match native popup style.

That's why I prefered to use scss instead of a bootstrap class. I can use another value

pparidans commented on addons/mail/static/src/core/web/messaging_menu.js:
```diff
@@ -80,6 +86,19 @@ export class MessagingMenu extends Component {
+    get installationRequest() {
+        return {
+            body: _t("Come here often? Install Odoo on your device!"),
+            displayName: sprintf(_t("%s has a suggestion"), this.store.odoobot.name),
```
I don't think you need to wrap the translation in `sprintf()` call ; just pass the argument to the `_t()` function.

Iucapad commented on addons/web/static/src/core/install_prompt/install_prompt_service.js:
```diff
@@ -0,0 +1,79 @@
+            browser.location.pathname !== "/web/tests" &&
+            (browser.BeforeInstallPromptEvent !== undefined ||
+                isIOS() ||
+                (isMacOS() && browser.navigator.userAgent.match(/Version\/(\d+)/)[1] >= 17));
```
I don't really know how we could detect that we are running on a version of Safari that has the 'Pin to dock' option available in the application menu, since there is no API related to that specifically

Iucapad replied:
To be more precise, it is only available on Safari 17 + on macOS Sonoma

pparidans replied:
I see your point.
On a side note, wouldn't `iOS() || isMacOS()` also match other browsers than Safari on these platforms ? :thinking: 
Should it also apply on those non-Safari browsers ? And if not, maybe `isBrowserSafari()` is more adequate ?

pparidans replied:
Are you sure about your logic for desktop Safari 17+ ?

Iucapad commented on addons/web/static/src/core/install_prompt/install_prompt_service.js:
```diff
@@ -0,0 +1,79 @@
+        // Finally, PWA are supported on desktop Safari since version 17
+        const canBeInstalled =
+            !isDisplayStandalone() &&
+            browser.location.pathname !== "/web/tests" &&
```
This was to avoid having to mess with mail tests with the notification area, which would fail on devices having the PWA installed, or if the browser don't support them (since assertions would fail about the number of notifications)

pparidans replied:
Not really sure about this... I'll invoke the high authority of sir @aab-odoo :grin: for a second opinion

pparidans commented on addons/web/static/src/core/install_prompt/install_prompt.scss:
```diff
@@ -0,0 +1,30 @@
+.o_install_prompt {
+    border-radius: 20px;
```
I've read your explanation about the `border-radius` value but, as it's an "unconventional" one, could you add the small comment about it ?

pparidans commented on addons/web/static/src/core/install_prompt/install_prompt.js:
```diff
@@ -0,0 +1,24 @@
+    };
+    static template = "web.InstallPrompt";
+
+    get sizeClass() {
+        return this.props.isMobileSafari && this.env.isSmall ? "mobile" : "desktop";
+    }
```
After re-reading it... Do we really need this ? Could usual media-queries being enough as the `InstallPrompt` is only triggered on Apple's OS (either macOS or iOS)

Iucapad replied:
Yes, why not :)

pparidans commented on addons/web/static/src/core/install_prompt/install_prompt_service.js:
```diff
@@ -0,0 +1,79 @@
+                const res = await nativePrompt.prompt();
+                browser.localStorage.setItem("pwa.installationState", res.outcome);
+                state.canPromptToInstall = false;
+            } else if (isIOS() || isMacOS()) {
```
Same as above

Iucapad replied:
Absolutely, I'll use `isBrowserSafari`

Iucapad commented:
Hello @alexkuhn could you check the second commit I added? This makes easier the patch of the service in the tests (as you can see in the 3 tests using patchInstallPromptService directly), and makes the install notification unavailable in each test using start or startServer

-> It removes the need to look and search '/tests' in the current URL to do this and the need to call the patch method for each test.

-> The reason I need this is that the notification appears as soon as the browser is able to install the app (depending support of PWA on the browser, if the app is running as PWA, if the notification is dismissed, or on apple OS). Many tests would fail when clicking on a notification or checking the count of unread notifications

alexkuhn commented:
@Iucapad It would be best to patch `browser` so that the install service exists in all tests but behaves how you want.
We could have a default behaviour of doing nothing (= default `patch` of browser in QUnit tests), and have `patchWithCleanup(browser, {})` before calls of `start()` in the specific tests you want this service to work the way you intent.

If you do this:
- all tests will use the genuine install service
- you won't have to mock a full service
- you don't have to make `start()` signature more complex than it currently is

pparidans reviewed (changes requested):
Slight comment about the desktop Safari 17+ detection.

Also, a rebase would be much needed (note: I removed the enterprise branch that was actually not used)

Iucapad commented:
@pparidans I just rebased and changed the safari detection, which was wrong in my use of isBrowserSafari (I don't have a better option about safari >= 17)

Iucapad commented:
Linked enterprise PR fixing patchWithCleanup of browser.navigator: https://github.com/odoo/enterprise/pull/48968

pparidans reviewed (approved):
@robodoo r+
